Spin dataset

therooler opened the pull request:
Sorry for taking so long with this follow up, I have been really swamped the last couple of weeks.
I implemented the `xxz_chain` function with the boilerplate code we wrote last time and added some tests. The new data can be found in the same repo as before: [XXZ_chain](https://github.com/therooler/tfq_data/tree/master/XXZ_chain). I still need to change the docstrings, but first I need to fix some other things.

I ran into a problem. For 4 and 8 qubits, I am getting a ground state fidelity close to 1 (perfect), but for 12 and 16, I get something ~ 0.3, even though when I created this data set with my own code, the residual energies and fidelities were perfect as well for these system sizes. I also ran a couple of tests with my own code, and I can consistenlty get ~1.0 fidelity and 1e-3 residual energies, so I don't think the model is the issue. These XXZ circuits are quite a bit deeper than the TFI ones, so maybe something is going wrong there, or I messed something up with the parameter mappings or circuit definition. I am a bit suspicious though since the N=4,8 qubit circuits run fine. 

I'll try to figure this out as soon as possible.

MichaelBroughton commented:
Are you on the latest version of tensorflow quantum with this data ?, we did have an issue like that that was fixed in the latest tensorflow quantum.

therooler commented:
Yes I am on the latest version. But I think I just figured out whats wrong; an error too embarrassing to make public. I'll get back to you asap.

therooler commented:
If guys could prepare the link for the XXZ data from my repo I will add the link in the code, and I expect that the library CI will pass.

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system.py:
```diff
@@ -320,3 +320,250 @@ def tfi_chain(qubits, boundary_condition="closed", data_dir=None):
+    circuit.append(cirq.H(qubits[i]) for i in range(0, nspins, 2))
+    circuit.append(
+        cirq.CNOT(qubits[j], qubits[j + 1]) for j in range(0, nspins - 1, 2))
+    for d in range(depth):
+        circuit.append(
+            cirq.ZZ(qubits[j], qubits[j + 1])**(symbols[d])
+            for j in range(1, nspins - 1, 2))
+        circuit.append(
+            cirq.YY(qubits[j], qubits[j + 1])**(symbols[d + depth])
+            for j in range(1, nspins - 1, 2))
+        circuit.append(
+            cirq.XX(qubits[j], qubits[j + 1])**(symbols[d + depth])
+            for j in range(1, nspins - 1, 2))
+        if boundary_condition == "closed":
+            circuit.append(cirq.ZZ(qubits[-1], qubits[0])**(symbols[d]))
+            circuit.append(cirq.YY(qubits[-1], qubits[0])**(symbols[d + depth]))
+            circuit.append(cirq.XX(qubits[-1], qubits[0])**(symbols[d + depth]))
+        circuit.append(
+            cirq.ZZ(qubits[j], qubits[j + 1])**(symbols[d + 2 * depth])
+            for j in range(0, nspins - 1, 2))
+        circuit.append(
+            cirq.YY(qubits[j], qubits[j + 1])**(symbols[d + 3 * depth])
+            for j in range(0, nspins - 1, 2))
+        circuit.append(
+            cirq.XX(qubits[j], qubits[j + 1])**(symbols[d + 3 * depth])
+            for j in range(0, nspins - 1, 2))
```
Repeating the `for j in range(1, nspins-1, 2)` loop over and over makes me wonder if we can't do something like:

```python
interacting_qubits = zip(qubits, qubits[1:])
if boundary_condition == 'closed':
  interacting_qubits.append((qubits[0], qubits[-1]))

for d in range(depth):
  for a, b in ineracting_qubits:
    circuit.append(cirq.XX(a, b))
    ....
    ....
    ....


```

therooler replied:
I tried this approach, but I am doing something wrong:
```python
even_qubits = zip([qubits[j] for j in range(0, nspins - 1, 2)],
                  [qubits[j + 1] for j in range(0, nspins - 1, 2)])
for q1,q2 in even_qubits:
   circuit.append(cirq.ZZ(q1,q2)**(symbols[d + 2 * depth]))
   circuit.append(cirq.YY(q1,q2)**(symbols[d + 3 * depth]))
   circuit.append cirq.XX(q1,q2)**(symbols[d + 3 * depth]))
```
gives the wrong fidelities and energies, whereas
```python
for j in range(0, nspins - 1, 2):
    circuit.append(cirq.ZZ(qubits[j], qubits[j + 1])**(symbols[d + 2 * depth]))
    circuit.append(cirq.YY(qubits[j], qubits[j + 1])**(symbols[d + 3 * depth]))
    circuit.append(cirq.XX(qubits[j], qubits[j + 1])**(symbols[d + 3 * depth]))
```
works correctly. What am I missing here?

MichaelBroughton replied:
I'm not at a computer to verify this, but I would guess that if the code breaks only when you substitute these two sections in and out that it has to do with the zip. I think it might be the fact that the zip state is producing a generator expression. Doing this:

```python3
even_qubits = list(zip(...))
```

Might fix things ?

therooler replied:
That did the trick.

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system.py:
```diff
@@ -320,3 +320,250 @@ def tfi_chain(qubits, boundary_condition="closed", data_dir=None):
+                                                 additional_info[i].params)
+        resolved_circuits.append(param_resolver)
+        # Make the PauliSum.
+        paulisum = sum(order_parameters[i] * cirq.Z(q1) * cirq.Z(q2)
+                       for q1, q2 in zip(qubits, qubits[1:]))
+        paulisum += sum(
+            cirq.Y(q1) * cirq.Y(q2) for q1, q2 in zip(qubits, qubits[1:]))
+        paulisum += sum(
+            cirq.X(q1) * cirq.X(q2) for q1, q2 in zip(qubits, qubits[1:]))
+        if boundary_condition == "closed":
+            paulisum += order_parameters[i] * cirq.Z(qubits[0]) * cirq.Z(
+                qubits[-1])
+            paulisum += cirq.Y(qubits[0]) * cirq.Y(qubits[-1])
+            paulisum += cirq.X(qubits[0]) * cirq.X(qubits[-1])
```
Can these also be made to use an `interacting_qubits` list like above ?

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system_test.py:
```diff
@@ -58,7 +58,7 @@ def test_fidelity(self):
                 phi = cirq.Simulator().simulate(circuits[n]).final_state
                 gs = addinfo[n].gs
-                self.assertAllClose(np.abs(np.conj(np.dot(gs, phi))),
+                self.assertAllClose(np.abs(np.dot(np.conj(gs), phi)),
```
try `np.vdot` instead ?

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system_test.py:
```diff
@@ -106,7 +106,105 @@ def test_param_resolver(self):
+            for n in range(len(addinfo)):
+                phi = cirq.Simulator().simulate(circuits[n]).final_state
+                gs = addinfo[n].gs
+                self.assertAllClose(np.abs(np.dot(np.conj(gs), phi)),
```
Try `np.vdot` ?

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system_test.py:
```diff
@@ -106,7 +106,105 @@ def test_param_resolver(self):
+                    circuits[n]).final_state
+                state_resolved_circuit = cirq.Simulator().simulate(
+                    resolved_circuit).final_state
+                self.assertAllClose(np.abs(
```
`np.vdot`

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system.py:
```diff
@@ -320,3 +320,250 @@ def tfi_chain(qubits, boundary_condition="closed", data_dir=None):
+    (3, 0): ───X───────X─────────────ZZ^-0.922──────────────YY^-0.915─── ...
+                           └──────────────────┘   └──────────────────┘
+
+    The labels indicate the phase of the system
```
Nit: space.

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system.py:
```diff
@@ -320,3 +320,250 @@ def tfi_chain(qubits, boundary_condition="closed", data_dir=None):
+    datapoint prepares (an approximation to) the ground state of the Hamiltonian
+    in the datapoint.
+
+    Example usage:
```
Nit: space.

therooler replied:
I feel dumb for asking this, but what do you mean by this? Add/remove a space?

MichaelBroughton replied:
Oh.... I weirdly thought you had an extra space in front of the E indenting it farther than everything else. Looking again it's toally fine hehe.

therooler replied:
Haha I see.

therooler commented:
I think if you could put the data in the google storage thing and give me the link we should be good.

MichaelBroughton commented:
Here we are: https://storage.googleapis.com/download.tensorflow.org/data/quantum/spin_systems/XXZ_chain.zip 
Another big thanks to @yashk2810 .

therooler commented:
Awesome, that works. But when I unpack the `TFI_chain.zip`, the folder is still called `TFI_chain-master` whereas the `XXZ_chain.zip` unpacks to `XXZ_chain`. Once that is changed, all tests should pass.

I'm on vacation for a couple more days, after that I will start with the 2D TFIM.

therooler commented:
It looks like the test takes too long for the CI to complete. Might be the 16 qubit circuits, which can take some time to load. Is there a way to make the tests more efficient using fixtures/namespaces or something? Then we could load the data only once for each system, and use only a single `spin_system.tfi_chain` and `spin_system.xxz_chain` call. I remember doing this at some point for another project.

MichaelBroughton commented:
Hmmm that is a long time, maybe instead of testing all circuits at all sizes, you could test a small random subset of circuits at all sizes ?

therooler commented:
I rewrote the tests to make sure that the data loading happens just once, but that did not resolve the timeout issue. I profiled some of the code outside of the test environment, and discovered that the recursive calls to `param_resolver` take a lot of time when loading the data: 0.5 sec in total per 12 qubit system for XXZ, which adds up if you do it 80 times. But it also seems that when I run stuff in the Bazel test environment, things slow down even more. 

I am also having a bit of trouble debugging the problem since I don't fully understand how Bazel testing + pytest works. The logs in `bazel-testlogs` provide some insight, but I don't know how to find out what part of the tests are slow...

Could we maybe schedule a meeting this week to discuss some of this stuff?

MichaelBroughton commented:
Oops, I missed this comment, maybe we should save this for this weeks sync meeting ?

therooler commented:
yeah that also works.

therooler commented:
So the good news is that library checks finally passed. The bad news is that it took 47 minutes.

MichaelBroughton commented:
Ahhh that is a while lol. Was that 47 minutes testing ALL of the circuits or a fractional subset ?

therooler commented:
10 random points for all 4,8,12 and 16 qubit systems.

therooler commented:
I timed the code for the 16 qubit xxz load function on my machine (Intel(R) Core(TM) i7-6700HQ ):

```python
from spin_system import xxz_chain, tfi_chain
import cirq
import numpy as np
import time

class XXZChainTest(object):
    """Testing tfi_chain."""

    def __init__(self, methodName):
        self.supported_nspins = [16]
        self.data_dict = {}
        self.qbs_dict = {}
        for nspins in self.supported_nspins:
            print(nspins)
            qbs = cirq.GridQubit.rect(nspins, 1)
            self.data_dict[nspins] = xxz_chain(
                qbs,
                'closed',
            )
            self.qbs_dict[nspins] = qbs
        self.random_subset = np.random.permutation(list(range(76)))[:10]

    def test_fidelity(self):
        """Test that all fidelities are close to 1."""
        for nspins in self.supported_nspins:
            circuits, _, _, addinfo = self.data_dict[nspins]
            for n in self.random_subset:
                phi = cirq.Simulator().simulate(circuits[n]).final_state
                gs = addinfo[n].gs

    def test_paulisum(self):
        """Test that the PauliSum Hamiltonians give the ground state energy."""
        for nspins in self.supported_nspins:
            circuits, _, pauli_sums, addinfo = self.data_dict[nspins]
            qubit_map = {self.qbs_dict[nspins][i]: i for i in range(nspins)}
            for n in self.random_subset:
                phi = cirq.Simulator().simulate(circuits[n]).final_state
                e = pauli_sums[n].expectation_from_wavefunction(phi, qubit_map)

    def test_param_resolver(self):
        """Test that the resolved circuits are correct."""
        for nspins in self.supported_nspins:
            circuits, _, _, addinfo = self.data_dict[nspins]
            for n in self.random_subset:
                resolved_circuit = cirq.resolve_parameters(
                    addinfo[n].var_circuit, addinfo[n].params)
                state_circuit = cirq.Simulator().simulate(
                    circuits[n]).final_state
                state_resolved_circuit = cirq.Simulator().simulate(
                    resolved_circuit).final_state
start = time.time()
test_class = XXZChainTest('duhdoi')
print("Class_init: {}".format(time.time()-start))
start = time.time()
test_class.test_fidelity()
print("Fidelity test: {}".format(time.time()-start))
start = time.time()
test_class.test_paulisum()
print("Paulisum test: {}".format(time.time()-start))
start = time.time()
test_class.test_param_resolver()
print("Param resolver: {}".format(time.time()-start))

```
Which produces:
```
/home/rooler/anaconda3/envs/tfq/bin/python /home/rooler/Documents/tfq/tensorflow_quantum/datasets/testing_stuff.py
2020-09-21 13:56:56.309162: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/lib64
2020-09-21 13:56:56.309187: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
16
Class_init: 76.4439525604248
Fidelity test: 5.773798227310181
Paulisum test: 6.041624307632446
Param resolver: 22.43285608291626

Process finished with exit code 0
```
So loading the data takes a long time, but after that it should be smooth sailing. The difference however is ~100 seconds versus 46 minutes between my laptop and the CI  setup. I tried to add the `tags = ["exclusive"]` to the BUILD file for the tests, to prevent Bazel from parallelizing this test, but it still takes 46 minutes. Do you have any other suggestion on how to troubleshoot this?

Also, I am preparing the code and data for the 2D TFI and XXZ models that we solved, so I intend to make another pull request for that this week.

therooler commented:
If I run the bazel tests locally the spin_data set takes about 10 minutes in total, which is still a lot compared to the 100 seconds that the 16 qubit xxz takes.

MichaelBroughton commented:
Hmmm if there is such a big disparity between the old dataset test and this one, I'm wondering if we might accidentally be loading the entire batch of wavefunctions from each file ? Perhaps the next step is to compare the numbers you are getting here to the numbers for tfi_chain and see which particular parts are so much slower in comparison.

therooler commented:
I see. Well, I think I narrowed down the problem in another way. I enabled logging to track what was happening during the bazel tests. For reasons that I can't quite understand, the `__init__` function I wrote that would supposedly load the data only once for each test suite, gets called 6 times (not sure why). This code (similar snippet for TFI) in `spin_systems_test.py`:
```python
class XXZChainTest(tf.test.TestCase):
    """Testing tfi_chain."""

    def __init__(self, method):
        self.supported_nspins = [4, 8, 12, 16]
        self.data_dict = {}
        self.qbs_dict = {}
        logging.warning('Looping here XXZ!')
        for nspins in self.supported_nspins:
            start = time.time()
            logging.warning('nspins={}'.format(nspins))
            qbs = cirq.GridQubit.rect(nspins, 1)
            self.data_dict[nspins] = spin_system.xxz_chain(
                qbs,
                'closed',
            )
            self.qbs_dict[nspins] = qbs
            logging.warning('Duration={}'.format(time.time() - start))
        self.random_subset = np.random.permutation(list(range(76)))[:10]
        super(XXZChainTest, self).__init__(method)
```
produces the `test.log`:
```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //tensorflow_quantum/datasets:spin_system_test
-----------------------------------------------------------------------------
2020-09-24 14:28:51.053731: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda-10.0/lib64:/usr/local/cuda/lib64
2020-09-24 14:28:51.053781: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
Running tests under Python 3.8.5: /home/rooler/anaconda3/envs/tfq/bin/python3
W0924 14:28:54.970267 140560489301760 spin_system_test.py:31] Looping here TFI!
W0924 14:28:54.970391 140560489301760 spin_system_test.py:34] nspins=4
W0924 14:28:59.964326 140560489301760 spin_system_test.py:40] Duration=4.993925094604492
W0924 14:28:59.964446 140560489301760 spin_system_test.py:34] nspins=8
W0924 14:29:04.555370 140560489301760 spin_system_test.py:40] Duration=4.590913534164429
W0924 14:29:04.555520 140560489301760 spin_system_test.py:34] nspins=12
W0924 14:29:09.271236 140560489301760 spin_system_test.py:40] Duration=4.715707540512085
W0924 14:29:09.271364 140560489301760 spin_system_test.py:34] nspins=16
W0924 14:29:14.415581 140560489301760 spin_system_test.py:40] Duration=5.144222259521484
W0924 14:29:14.415829 140560489301760 spin_system_test.py:31] Looping here TFI!
W0924 14:29:14.415909 140560489301760 spin_system_test.py:34] nspins=4
W0924 14:29:18.602371 140560489301760 spin_system_test.py:40] Duration=4.18643593788147
W0924 14:29:18.602505 140560489301760 spin_system_test.py:34] nspins=8
W0924 14:29:22.915986 140560489301760 spin_system_test.py:40] Duration=4.313488960266113
W0924 14:29:22.916116 140560489301760 spin_system_test.py:34] nspins=12
W0924 14:29:27.787804 140560489301760 spin_system_test.py:40] Duration=4.871696472167969
W0924 14:29:27.787934 140560489301760 spin_system_test.py:34] nspins=16
W0924 14:29:32.779796 140560489301760 spin_system_test.py:40] Duration=4.991867303848267
W0924 14:29:32.780022 140560489301760 spin_system_test.py:31] Looping here TFI!
W0924 14:29:32.780094 140560489301760 spin_system_test.py:34] nspins=4
W0924 14:29:37.398508 140560489301760 spin_system_test.py:40] Duration=4.618398904800415
W0924 14:29:37.398650 140560489301760 spin_system_test.py:34] nspins=8
W0924 14:29:41.756363 140560489301760 spin_system_test.py:40] Duration=4.357707262039185
W0924 14:29:41.756504 140560489301760 spin_system_test.py:34] nspins=12
W0924 14:29:46.373649 140560489301760 spin_system_test.py:40] Duration=4.61715292930603
W0924 14:29:46.373791 140560489301760 spin_system_test.py:34] nspins=16
W0924 14:29:51.697457 140560489301760 spin_system_test.py:40] Duration=5.323667287826538
W0924 14:29:51.697691 140560489301760 spin_system_test.py:31] Looping here TFI!
W0924 14:29:51.697777 140560489301760 spin_system_test.py:34] nspins=4
W0924 14:29:56.287828 140560489301760 spin_system_test.py:40] Duration=4.590041637420654
W0924 14:29:56.287946 140560489301760 spin_system_test.py:34] nspins=8
W0924 14:30:00.942666 140560489301760 spin_system_test.py:40] Duration=4.654702186584473
W0924 14:30:00.942839 140560489301760 spin_system_test.py:34] nspins=12
W0924 14:30:05.866748 140560489301760 spin_system_test.py:40] Duration=4.923896551132202
W0924 14:30:05.866864 140560489301760 spin_system_test.py:34] nspins=16
W0924 14:30:11.339686 140560489301760 spin_system_test.py:40] Duration=5.472797155380249
W0924 14:30:11.339947 140560489301760 spin_system_test.py:31] Looping here TFI!
W0924 14:30:11.340031 140560489301760 spin_system_test.py:34] nspins=4
W0924 14:30:15.625852 140560489301760 spin_system_test.py:40] Duration=4.285811185836792
W0924 14:30:15.625959 140560489301760 spin_system_test.py:34] nspins=8
W0924 14:30:20.823881 140560489301760 spin_system_test.py:40] Duration=5.19791054725647
W0924 14:30:20.824007 140560489301760 spin_system_test.py:34] nspins=12
W0924 14:30:25.961658 140560489301760 spin_system_test.py:40] Duration=5.137648582458496
W0924 14:30:25.961802 140560489301760 spin_system_test.py:34] nspins=16
W0924 14:30:31.070449 140560489301760 spin_system_test.py:40] Duration=5.108652353286743
W0924 14:30:31.070665 140560489301760 spin_system_test.py:31] Looping here TFI!
W0924 14:30:31.070732 140560489301760 spin_system_test.py:34] nspins=4
W0924 14:30:35.333079 140560489301760 spin_system_test.py:40] Duration=4.262335538864136
W0924 14:30:35.333209 140560489301760 spin_system_test.py:34] nspins=8
W0924 14:30:40.187439 140560489301760 spin_system_test.py:40] Duration=4.854228734970093
W0924 14:30:40.187694 140560489301760 spin_system_test.py:34] nspins=12
W0924 14:30:45.630519 140560489301760 spin_system_test.py:40] Duration=5.442853212356567
W0924 14:30:45.630682 140560489301760 spin_system_test.py:34] nspins=16
W0924 14:30:51.929644 140560489301760 spin_system_test.py:40] Duration=6.298946142196655
W0924 14:30:51.931608 140560489301760 spin_system_test.py:125] Looping here XXZ!
W0924 14:30:51.931681 140560489301760 spin_system_test.py:129] nspins=4
W0924 14:30:56.171225 140560489301760 spin_system_test.py:136] Duration=4.239532709121704
W0924 14:30:56.171367 140560489301760 spin_system_test.py:129] nspins=8
W0924 14:31:07.740903 140560489301760 spin_system_test.py:136] Duration=11.569547414779663
W0924 14:31:07.741045 140560489301760 spin_system_test.py:129] nspins=12
W0924 14:31:42.736577 140560489301760 spin_system_test.py:136] Duration=34.995537757873535
W0924 14:31:42.736701 140560489301760 spin_system_test.py:129] nspins=16
W0924 14:32:59.099303 140560489301760 spin_system_test.py:136] Duration=76.36258220672607
W0924 14:32:59.099803 140560489301760 spin_system_test.py:125] Looping here XXZ!
W0924 14:32:59.099887 140560489301760 spin_system_test.py:129] nspins=4
W0924 14:33:03.942298 140560489301760 spin_system_test.py:136] Duration=4.842375993728638
W0924 14:33:03.942595 140560489301760 spin_system_test.py:129] nspins=8
W0924 14:33:16.663506 140560489301760 spin_system_test.py:136] Duration=12.720907926559448
W0924 14:33:16.663667 140560489301760 spin_system_test.py:129] nspins=12
W0924 14:33:52.335239 140560489301760 spin_system_test.py:136] Duration=35.671563148498535
W0924 14:33:52.337006 140560489301760 spin_system_test.py:129] nspins=16
W0924 14:35:04.349862 140560489301760 spin_system_test.py:136] Duration=72.01285457611084
W0924 14:35:04.354977 140560489301760 spin_system_test.py:125] Looping here XXZ!
W0924 14:35:04.355069 140560489301760 spin_system_test.py:129] nspins=4
W0924 14:35:09.585326 140560489301760 spin_system_test.py:136] Duration=5.230235576629639
W0924 14:35:09.585494 140560489301760 spin_system_test.py:129] nspins=8
W0924 14:35:23.444278 140560489301760 spin_system_test.py:136] Duration=13.858757734298706
W0924 14:35:23.444530 140560489301760 spin_system_test.py:129] nspins=12
W0924 14:35:59.913906 140560489301760 spin_system_test.py:136] Duration=36.46936821937561
W0924 14:35:59.914056 140560489301760 spin_system_test.py:129] nspins=16
W0924 14:37:16.357321 140560489301760 spin_system_test.py:136] Duration=76.44327044487
W0924 14:37:16.365069 140560489301760 spin_system_test.py:125] Looping here XXZ!
W0924 14:37:16.365227 140560489301760 spin_system_test.py:129] nspins=4
W0924 14:37:21.389653 140560489301760 spin_system_test.py:136] Duration=5.02441668510437
W0924 14:37:21.389801 140560489301760 spin_system_test.py:129] nspins=8
W0924 14:37:35.698544 140560489301760 spin_system_test.py:136] Duration=14.308732032775879
W0924 14:37:35.698687 140560489301760 spin_system_test.py:129] nspins=12
W0924 14:38:12.186959 140560489301760 spin_system_test.py:136] Duration=36.488274574279785
W0924 14:38:12.187133 140560489301760 spin_system_test.py:129] nspins=16
W0924 14:39:28.510338 140560489301760 spin_system_test.py:136] Duration=76.32319211959839
W0924 14:39:28.524270 140560489301760 spin_system_test.py:125] Looping here XXZ!
W0924 14:39:28.524366 140560489301760 spin_system_test.py:129] nspins=4
W0924 14:39:33.505145 140560489301760 spin_system_test.py:136] Duration=4.980764150619507
W0924 14:39:33.513297 140560489301760 spin_system_test.py:129] nspins=8
W0924 14:39:48.592314 140560489301760 spin_system_test.py:136] Duration=15.07901906967163
W0924 14:39:48.598765 140560489301760 spin_system_test.py:129] nspins=12
W0924 14:40:25.861888 140560489301760 spin_system_test.py:136] Duration=37.26315355300903
W0924 14:40:25.865570 140560489301760 spin_system_test.py:129] nspins=16
W0924 14:41:45.738578 140560489301760 spin_system_test.py:136] Duration=79.87084937095642
W0924 14:41:45.743028 140560489301760 spin_system_test.py:125] Looping here XXZ!
W0924 14:41:45.743176 140560489301760 spin_system_test.py:129] nspins=4
W0924 14:41:50.872711 140560489301760 spin_system_test.py:136] Duration=5.129522323608398
W0924 14:41:50.877364 140560489301760 spin_system_test.py:129] nspins=8
W0924 14:42:05.361292 140560489301760 spin_system_test.py:136] Duration=14.483922004699707
W0924 14:42:05.377101 140560489301760 spin_system_test.py:129] nspins=12
W0924 14:42:42.285480 140560489301760 spin_system_test.py:136] Duration=36.90791201591492
W0924 14:42:42.287755 140560489301760 spin_system_test.py:129] nspins=16
W0924 14:43:53.558661 140560489301760 spin_system_test.py:136] Duration=71.27091193199158
[ RUN      ] TFIChainTest.test_errors
INFO:tensorflow:time(__main__.TFIChainTest.test_errors): 0.0s
I0924 14:43:53.564625 140560489301760 test_util.py:1972] time(__main__.TFIChainTest.test_errors): 0.0s
[       OK ] TFIChainTest.test_errors
[ RUN      ] TFIChainTest.test_fidelity
INFO:tensorflow:time(__main__.TFIChainTest.test_fidelity): 1.7s
I0924 14:43:55.261113 140560489301760 test_util.py:1972] time(__main__.TFIChainTest.test_fidelity): 1.7s
[       OK ] TFIChainTest.test_fidelity
[ RUN      ] TFIChainTest.test_param_resolver
INFO:tensorflow:time(__main__.TFIChainTest.test_param_resolver): 3.38s
I0924 14:43:58.643907 140560489301760 test_util.py:1972] time(__main__.TFIChainTest.test_param_resolver): 3.38s
[       OK ] TFIChainTest.test_param_resolver
[ RUN      ] TFIChainTest.test_paulisum
INFO:tensorflow:time(__main__.TFIChainTest.test_paulisum): 1.86s
I0924 14:44:00.503697 140560489301760 test_util.py:1972] time(__main__.TFIChainTest.test_paulisum): 1.86s
[       OK ] TFIChainTest.test_paulisum
[ RUN      ] TFIChainTest.test_returned_objects
INFO:tensorflow:time(__main__.TFIChainTest.test_returned_objects): 0.0s
I0924 14:44:00.504518 140560489301760 test_util.py:1972] time(__main__.TFIChainTest.test_returned_objects): 0.0s
[       OK ] TFIChainTest.test_returned_objects
[ RUN      ] TFIChainTest.test_session
[  SKIPPED ] TFIChainTest.test_session
[ RUN      ] XXZChainTest.test_errors
INFO:tensorflow:time(__main__.XXZChainTest.test_errors): 0.0s
I0924 14:44:00.504962 140560489301760 test_util.py:1972] time(__main__.XXZChainTest.test_errors): 0.0s
[       OK ] XXZChainTest.test_errors
[ RUN      ] XXZChainTest.test_fidelity
INFO:tensorflow:time(__main__.XXZChainTest.test_fidelity): 5.76s
I0924 14:44:06.262632 140560489301760 test_util.py:1972] time(__main__.XXZChainTest.test_fidelity): 5.76s
[       OK ] XXZChainTest.test_fidelity
[ RUN      ] XXZChainTest.test_param_resolver
INFO:tensorflow:time(__main__.XXZChainTest.test_param_resolver): 25.31s
I0924 14:44:31.569550 140560489301760 test_util.py:1972] time(__main__.XXZChainTest.test_param_resolver): 25.31s
[       OK ] XXZChainTest.test_param_resolver
[ RUN      ] XXZChainTest.test_paulisum
INFO:tensorflow:time(__main__.XXZChainTest.test_paulisum): 6.12s
I0924 14:44:37.686982 140560489301760 test_util.py:1972] time(__main__.XXZChainTest.test_paulisum): 6.12s
[       OK ] XXZChainTest.test_paulisum
[ RUN      ] XXZChainTest.test_returned_objects
INFO:tensorflow:time(__main__.XXZChainTest.test_returned_objects): 0.0s
I0924 14:44:37.687764 140560489301760 test_util.py:1972] time(__main__.XXZChainTest.test_returned_objects): 0.0s
[       OK ] XXZChainTest.test_returned_objects
[ RUN      ] XXZChainTest.test_session
[  SKIPPED ] XXZChainTest.test_session
----------------------------------------------------------------------
Ran 12 tests in 44.127s

OK (skipped=2)

```
So the actual tests only take 40 seconds. Working on a fix atm.

MichaelBroughton commented:
Thanks for investigating, can we try:
1. Cirq 0.8.2.
2. Cirq-unstable. 
3. Failing all of that open an issue on Cirq

therooler commented:
This is with cirq 0.8.2:
Code that's being timed both TFI and XXZ:
```python
...
    for i, directory in enumerate(x for x in os.listdir(data_path)):
        total_start = time.time()

        # The folders are named according to the order value data they contain.
        g = float(directory)
        with open(os.path.join(data_path, directory, "stats.txt"), "r") as file:
            lines = file.readlines()
            res_e = float(lines[0].split("=")[1].strip("\n"))
            fidelity = float(lines[2].split("=")[1].strip("\n"))
        order_parameters.append(g)
        params = np.load(os.path.join(data_path, directory, "params.npy")) \
                 / np.pi
        additional_info.append(
            SpinSystemInfo(g=g,
                           gs=np.load(
                               os.path.join(data_path, directory,
                                            "groundstate.npy"))[:, 0],
                           gs_energy=np.load(
                               os.path.join(data_path, directory,
                                            "energy.npy"))[0],
                           res_energy=res_e,
                           fidelity=fidelity,
                           params=dict(zip(symbol_names, params.flatten())),
                           var_circuit=circuit))
        # Resolve the circuit parameters.
        resolve_start = time.time()
        param_resolver = cirq.resolve_parameters(circuit,
                                                 additional_info[i].params)
        t_param += time.time()-resolve_start
        resolved_circuits.append(param_resolver)
        # Make the PauliSum.
        paulisum = sum(order_parameters[i] * cirq.Z(q1) * cirq.Z(q2)
                       for q1, q2 in zip(qubits, qubits[1:]))
        paulisum += sum(
            cirq.Y(q1) * cirq.Y(q2) for q1, q2 in zip(qubits, qubits[1:]))
        paulisum += sum(
            cirq.X(q1) * cirq.X(q2) for q1, q2 in zip(qubits, qubits[1:]))
        if boundary_condition == "closed":
            paulisum += order_parameters[i] * cirq.Z(qubits[0]) * cirq.Z(
                qubits[-1])
            paulisum += cirq.Y(qubits[0]) * cirq.Y(qubits[-1])
            paulisum += cirq.X(qubits[0]) * cirq.X(qubits[-1])
        hamiltonians.append(paulisum)

        # Set labels for the different phases.
        if order_parameters[i] <= 1.0:
            labels.append(0)
        else:
            labels.append(1)
        t_total+= time.time()-total_start
    logging.warning('Total time: {}'.format(t_total))
    logging.warning('Total param resolve time: {}'.format(t_param))
    logging.warning('Total other time: {}'.format(t_total-t_param))
...
```
With driver:
```python
...
    # XXZ CHAIN
    SUPPORTED_NSPINS_XXZ_CHAIN = [4, 8, 12, 16]
    DATA_DICT_XXZ_CHAIN = {}
    QBS_DICT_XXZ_CHAIN = {}
    for nspins in SUPPORTED_NSPINS_XXZ_CHAIN:
        QBS_XXZ_CHAIN = cirq.GridQubit.rect(nspins, 1)
        logging.warning('\nCalling xxz_chain, nspins={}\n'.format(nspins))
        start = time.time()
        DATA_DICT_XXZ_CHAIN[nspins] = spin_system.xxz_chain(
            QBS_XXZ_CHAIN,
            'closed',
        )
        logging.warning('\nEnd of function call, duration {}\n'.format(time.time()-start))

        QBS_DICT_XXZ_CHAIN[nspins] = QBS_XXZ_CHAIN
    RANDOM_SUBSET_XXZ_CHAIN = np.random.permutation(list(
        range(76)))[:RANDOM_SUBSET_SIZE]

    tf.test.main()
...
```
produces test.log in `bazel-testlogs`

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //tensorflow_quantum/datasets:spin_system_test
-----------------------------------------------------------------------------
CUDA complaints
WARNING:root:
Calling tfi_chain, nspins=4

WARNING:root:Total time: 0.17909455299377441
WARNING:root:Total param resolve time: 0.03914499282836914
WARNING:root:Total other time: 0.13994956016540527
WARNING:root:
End of function call, duration 4.698606967926025

WARNING:root:
Calling tfi_chain, nspins=8

WARNING:root:Total time: 0.43994903564453125
WARNING:root:Total param resolve time: 0.1330249309539795
WARNING:root:Total other time: 0.30692410469055176
WARNING:root:
End of function call, duration 4.9622962474823

WARNING:root:
Calling tfi_chain, nspins=12

WARNING:root:Total time: 0.7621908187866211
WARNING:root:Total param resolve time: 0.3138411045074463
WARNING:root:Total other time: 0.4483497142791748
WARNING:root:
End of function call, duration 4.555153131484985

WARNING:root:
Calling tfi_chain, nspins=16

WARNING:root:Total time: 1.0340774059295654
WARNING:root:Total param resolve time: 0.613628625869751
WARNING:root:Total other time: 0.42044878005981445
WARNING:root:
End of function call, duration 5.729584455490112

WARNING:root:
Calling xxz_chain, nspins=4

WARNING:root:Total time: 1.7049949169158936
WARNING:root:Total param resolve time: 1.4906489849090576
WARNING:root:Total other time: 0.21434593200683594
WARNING:root:
End of function call, duration 4.571383953094482

WARNING:root:
Calling xxz_chain, nspins=8

WARNING:root:Total time: 10.469760179519653
WARNING:root:Total param resolve time: 9.955015659332275
WARNING:root:Total other time: 0.5147445201873779
WARNING:root:
End of function call, duration 12.793200492858887

WARNING:root:
Calling xxz_chain, nspins=12

WARNING:root:Total time: 32.966421127319336
WARNING:root:Total param resolve time: 32.47918367385864
WARNING:root:Total other time: 0.48723745346069336
WARNING:root:
End of function call, duration 35.31800150871277

WARNING:root:
Calling xxz_chain, nspins=16

WARNING:root:Total time: 77.75785803794861
WARNING:root:Total param resolve time: 77.08113980293274
WARNING:root:Total other time: 0.6767182350158691
WARNING:root:
End of function call, duration 80.02726912498474

Running tests under Python 3.8.5: /home/rooler/anaconda3/envs/tfq/bin/python3
[ RUN      ] TFIChainTest.test_errors
INFO:tensorflow:time(__main__.TFIChainTest.test_errors): 0.0s
I0924 18:30:17.449815 140645984651008 test_util.py:1972] time(__main__.TFIChainTest.test_errors): 0.0s
[       OK ] TFIChainTest.test_errors
[ RUN      ] TFIChainTest.test_fidelity
INFO:tensorflow:time(__main__.TFIChainTest.test_fidelity): 1.86s
I0924 18:30:19.306783 140645984651008 test_util.py:1972] time(__main__.TFIChainTest.test_fidelity): 1.86s
[       OK ] TFIChainTest.test_fidelity
[ RUN      ] TFIChainTest.test_param_resolver
INFO:tensorflow:time(__main__.TFIChainTest.test_param_resolver): 3.74s
I0924 18:30:23.054289 140645984651008 test_util.py:1972] time(__main__.TFIChainTest.test_param_resolver): 3.74s
[       OK ] TFIChainTest.test_param_resolver
[ RUN      ] TFIChainTest.test_paulisum
INFO:tensorflow:time(__main__.TFIChainTest.test_paulisum): 2.05s
I0924 18:30:25.102919 140645984651008 test_util.py:1972] time(__main__.TFIChainTest.test_paulisum): 2.05s
[       OK ] TFIChainTest.test_paulisum
[ RUN      ] TFIChainTest.test_returned_objects
INFO:tensorflow:time(__main__.TFIChainTest.test_returned_objects): 0.0s
I0924 18:30:25.103800 140645984651008 test_util.py:1972] time(__main__.TFIChainTest.test_returned_objects): 0.0s
[       OK ] TFIChainTest.test_returned_objects
[ RUN      ] TFIChainTest.test_session
[  SKIPPED ] TFIChainTest.test_session
[ RUN      ] XXZChainTest.test_errors
INFO:tensorflow:time(__main__.XXZChainTest.test_errors): 0.0s
I0924 18:30:25.104405 140645984651008 test_util.py:1972] time(__main__.XXZChainTest.test_errors): 0.0s
[       OK ] XXZChainTest.test_errors
[ RUN      ] XXZChainTest.test_fidelity
INFO:tensorflow:time(__main__.XXZChainTest.test_fidelity): 7.4s
I0924 18:30:32.503252 140645984651008 test_util.py:1972] time(__main__.XXZChainTest.test_fidelity): 7.4s
[       OK ] XXZChainTest.test_fidelity
[ RUN      ] XXZChainTest.test_param_resolver
INFO:tensorflow:time(__main__.XXZChainTest.test_param_resolver): 29.45s
I0924 18:31:01.950268 140645984651008 test_util.py:1972] time(__main__.XXZChainTest.test_param_resolver): 29.45s
[       OK ] XXZChainTest.test_param_resolver
[ RUN      ] XXZChainTest.test_paulisum
INFO:tensorflow:time(__main__.XXZChainTest.test_paulisum): 6.99s
I0924 18:31:08.943530 140645984651008 test_util.py:1972] time(__main__.XXZChainTest.test_paulisum): 6.99s
[       OK ] XXZChainTest.test_paulisum
[ RUN      ] XXZChainTest.test_returned_objects
INFO:tensorflow:time(__main__.XXZChainTest.test_returned_objects): 0.0s
I0924 18:31:08.944404 140645984651008 test_util.py:1972] time(__main__.XXZChainTest.test_returned_objects): 0.0s
[       OK ] XXZChainTest.test_returned_objects
[ RUN      ] XXZChainTest.test_session
[  SKIPPED ] XXZChainTest.test_session
----------------------------------------------------------------------
Ran 12 tests in 51.496s

OK (skipped=2)
```
So the param resolver seems to be the main problem.

therooler commented:
I discovered the issue. The problem was that the parameters were being loaded as `np.float32`s, and for some reason, `cirq.parameter_resolver` really does not like that. I opened an issue on the Cirq git

All is good for this PR now, if you think the code looks good, we can finally merge!

MichaelBroughton commented:
Ahhh awesome detective work! What sorts of times were you getting after changing from `np.float32` ?

therooler commented:
Down to 72 seconds on the CI. That's more like it.

MichaelBroughton commented on tensorflow_quantum/datasets/BUILD:
```diff
@@ -27,6 +27,7 @@ py_test(
 py_test(
     name = "spin_system_test",
     srcs = ["spin_system_test.py"],
+    timeout = "eternal",
```
Is this still needed ?

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system.py:
```diff
@@ -312,6 +312,248 @@ def tfi_chain(qubits, boundary_condition="closed", data_dir=None):
             labels.append(1)
         else:
             labels.append(2)
+    # Make sure that the data is ordered from g=0.2 to g=1.8.
+    _, resolved_circuits, labels, hamiltonians, additional_info = zip(*sorted(
+        zip(order_parameters, resolved_circuits, labels, hamiltonians,
+            additional_info)))
+
+    return resolved_circuits, labels, hamiltonians, additional_info
```
Was there a change made to `tfi_chain` os this the diff acting weird ?

therooler replied:
No changes here. Don't know why this shows up.

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system.py:
```diff
@@ -312,6 +312,248 @@ def tfi_chain(qubits, boundary_condition="closed", data_dir=None):
+
+
+def xxz_chain(qubits, boundary_condition="closed", data_dir=None):
+    """XXZ model quantum data set.
```
Should we mention that this is 1D ?

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system.py:
```diff
@@ -312,6 +312,248 @@ def tfi_chain(qubits, boundary_condition="closed", data_dir=None):
+    {'theta_0': 1.0780547, 'theta_1': 0.99271035, 'theta_2': 1.0854135, ...
+
+    and change them to experiment with different parameter values by using
+    the unresolved variational circuit returned by tfichain
```
XXZ ?

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system.py:
```diff
@@ -312,6 +312,248 @@ def tfi_chain(qubits, boundary_condition="closed", data_dir=None):
+    # Load the data and append to the lists.
+    for i, directory in enumerate(x for x in os.listdir(data_path)):
+        # The folders are named according to the order value data they contain.
+        g = float(directory)
+        with open(os.path.join(data_path, directory, "stats.txt"), "r") as file:
+            lines = file.readlines()
+            res_e = float(lines[0].split("=")[1].strip("\n"))
+            fidelity = float(lines[2].split("=")[1].strip("\n"))
+        order_parameters.append(g)
+        params = np.load(os.path.join(data_path, directory, "params.npy")) \
+                 / np.pi
+        params = params.astype(np.float64)
+        additional_info.append(
+            SpinSystemInfo(g=g,
+                           gs=np.load(
+                               os.path.join(data_path, directory,
+                                            "groundstate.npy"))[:, 0],
+                           gs_energy=np.load(
+                               os.path.join(data_path, directory,
+                                            "energy.npy"))[0],
+                           res_energy=res_e,
+                           fidelity=fidelity,
+                           params=dict(zip(symbol_names, params.flatten())),
+                           var_circuit=circuit))
+        # Resolve the circuit parameters.
+        param_resolver = cirq.resolve_parameters(circuit,
+                                                 additional_info[i].params)
+        resolved_circuits.append(param_resolver)
```
Would you mind referencing your Cirq Bug in a comment in the code so we know to revisit it later and why there is a cast of numeric types ?

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system.py:
```diff
@@ -312,6 +312,248 @@ def tfi_chain(qubits, boundary_condition="closed", data_dir=None):
+                           params=dict(zip(symbol_names, params.flatten())),
+                           var_circuit=circuit))
+        # Resolve the circuit parameters.
+        param_resolver = cirq.resolve_parameters(circuit,
```
Might want to call this `resolved_circuit`

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system.py:
```diff
@@ -312,6 +312,248 @@ def tfi_chain(qubits, boundary_condition="closed", data_dir=None):
+                                                 additional_info[i].params)
+        resolved_circuits.append(param_resolver)
+        # Make the PauliSum.
+        paulisum = sum(order_parameters[i] * cirq.Z(q1) * cirq.Z(q2)
+                       for q1, q2 in zip(qubits, qubits[1:]))
+        paulisum += sum(
+            cirq.Y(q1) * cirq.Y(q2) for q1, q2 in zip(qubits, qubits[1:]))
+        paulisum += sum(
+            cirq.X(q1) * cirq.X(q2) for q1, q2 in zip(qubits, qubits[1:]))
```
Does this accomplish the same thing (sorry for weird git code formats) ?

```python3
paulisum = sum(order_parameters[i] * cirq.Z(q1) * cirq.Z(q2) + \
               cirq.Y(q1) * cirq.Y(q2) + \
               cirq.X(q1) * cirq.X(q2)
                       for q1, q2 in zip(qubits, qubits[1:]))

```

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system.py:
```diff
@@ -312,6 +312,248 @@ def tfi_chain(qubits, boundary_condition="closed", data_dir=None):
+
+    # Define the circuit.
+    circuit = cirq.Circuit(cirq.X.on_each(qubits))
+    circuit.append(cirq.H(qubits[i]) for i in range(0, nspins, 2))
+    circuit.append(
+        cirq.CNOT(qubits[j], qubits[j + 1]) for j in range(0, nspins - 1, 2))
+
+    odd_qubits = list(
+        zip([qubits[j] for j in range(1, nspins - 1, 2)],
+            [qubits[j + 1] for j in range(1, nspins - 1, 2)]))
+    even_qubits = list(
+        zip([qubits[j] for j in range(0, nspins - 1, 2)],
+            [qubits[j + 1] for j in range(0, nspins - 1, 2)]))
```
The interchange of zip statements and slightly different indices is kind of confusing here. Might make a lot more sense if when defining the circuit we did something like:

```python3
even_qubits = qubits[::2]
odd_qubits = qubits[1::2]

circuit.append(cirq.H.on_each(even_qubits))
circuit.append(cirq.CNOT(e, o) for e, o in zip(even_qubits, odd_qubits))
.
.
.

```

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system.py:
```diff
@@ -312,6 +312,248 @@ def tfi_chain(qubits, boundary_condition="closed", data_dir=None):
+        paulisum += sum(
+            cirq.X(q1) * cirq.X(q2) for q1, q2 in zip(qubits, qubits[1:]))
+        if boundary_condition == "closed":
+            paulisum += order_parameters[i] * cirq.Z(qubits[0]) * cirq.Z(
+                qubits[-1])
+            paulisum += cirq.Y(qubits[0]) * cirq.Y(qubits[-1])
+            paulisum += cirq.X(qubits[0]) * cirq.X(qubits[-1])
```
Lets add these in one expression

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system_test.py:
```diff
@@ -93,23 +82,126 @@ def test_returned_objects(self):
 
 
 if __name__ == '__main__':
+    RANDOM_SUBSET_SIZE = 10
+    # TFI CHAIN
+    SUPPORTED_NSPINS_TFI_CHAIN = [4, 8, 12, 16]
+    DATA_DICT_TFI_CHAIN = {}
+    QBS_DICT_TFI_CHAIN = {}
+    for nspins in SUPPORTED_NSPINS_TFI_CHAIN:
+        QBS_TFI_CHAIN = cirq.GridQubit.rect(nspins, 1)
+        DATA_DICT_TFI_CHAIN[nspins] = spin_system.tfi_chain(
+            QBS_TFI_CHAIN,
+            'closed',
+        )
+        QBS_DICT_TFI_CHAIN[nspins] = QBS_TFI_CHAIN
+    RANDOM_SUBSET_TFI_CHAIN = np.random.permutation(list(
+        range(76)))[:RANDOM_SUBSET_SIZE]
+    # XXZ CHAIN
+    SUPPORTED_NSPINS_XXZ_CHAIN = [4, 8, 12, 16]
+    DATA_DICT_XXZ_CHAIN = {}
+    QBS_DICT_XXZ_CHAIN = {}
+    for nspins in SUPPORTED_NSPINS_XXZ_CHAIN:
+        QBS_XXZ_CHAIN = cirq.GridQubit.rect(nspins, 1)
+        DATA_DICT_XXZ_CHAIN[nspins] = spin_system.xxz_chain(
+            QBS_XXZ_CHAIN,
+            'closed',
+        )
+        QBS_DICT_XXZ_CHAIN[nspins] = QBS_XXZ_CHAIN
+    RANDOM_SUBSET_XXZ_CHAIN = np.random.permutation(list(
+        range(76)))[:RANDOM_SUBSET_SIZE]
```
Can we remove this so we just call `tf.test.main()` ?

therooler replied:
I moved this to below the imports. I can't figure out another way to make sure that the data loader is called only once when testing.

MichaelBroughton replied:
Would something like this work ? 

https://www.tensorflow.org/api_docs/python/tf/test/TestCase#setUpClass

therooler replied:
Yes, but then pylint complains that the method names are not correct.

therooler replied:
I added an exception for pylint to ignore method names for both tests.

MichaelBroughton reviewed (changes requested):
Looking good! I Think after this round of review we will be ready to merge.

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system_test.py:
```diff
@@ -93,23 +101,120 @@ def test_returned_objects(self):
+    # pylint: disable=C0103
+
+    @classmethod
+    def setUpClass(self):
```
Can you be sure to include a call to the superclasses `setUpClass` function as well ?

MichaelBroughton commented on tensorflow_quantum/datasets/spin_system_test.py:
```diff
@@ -22,6 +22,24 @@
+    # pylint: disable=C0103
+
+    @classmethod
+    def setUpClass(self):
```
Same here.

MichaelBroughton reviewed (commented):
LGTM
